File: leetcode/c++/15 3Sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int>& nums) {
        vector<vector<int> > ret;
        vector<int > vtemp;
        int len = nums.size();
        sort(nums.begin(),nums.end());//sort the input
        for(int i=0;i<len-2;i++){
            //find the tripe for each nums[i]
            // j1 and j2 log the index of the other two numbers
            if(i ==0 ||(i>0 && nums[i] != nums[i-1])){
                int p1 = i+1, p2 = len-1; // set two pointers
                while(p1 < p2){
                    if(nums[p1] + nums[p2] < -nums[i]){
                        p1++;
                    }else if(nums[p1] + nums[p2] == -nums[i]){
                        if(p1 == i+1){
                            vector<int > vtemp{nums[i], nums[p1], nums[p2]};
                            ret.push_back(vtemp);
                            vtemp.clear();

                        }else if(nums[p1] != nums[p1-1]){
                            vector<int > vtemp{nums[i], nums[p1], nums[p2]};
                            ret.push_back(vtemp);
                            vtemp.clear();

                        }
                        p1++,p2--;
                    }else{
                         p2--;
                    }
                }
            }


        }
        return ret;

    }
};

File: leetcode/go/15 3Sum.go
func helper(nums *[]int, res *[][]int){
    for k, v := range(*nums){
        if(v > 0){
            return
        }
        
        if(k > len(*nums) - 3){
            return
        }
        
        if(k > 0 && (*nums)[k] == (*nums)[k - 1]){
            continue
        }
        
        left := k + 1
        right := len(*nums) - 1
        
        for;left < right;{
            if(v + (*nums)[left] + (*nums)[right] > 0){
                right--
            }else if(v + (*nums)[left] + (*nums)[right] < 0){
                left++
            }else{
                *res = append(*res, []int{v, (*nums)[left], (*nums)[right]})
                left++
                right--
                
                for;left < right;{
                    if(left != 0 && (*nums)[left] == (*nums)[left - 1]){
                        left++
                    }else{
                        break
                    }
                }
            
                for;left < right;{
                    if(right != len(*nums) - 1 && (*nums)[right] == (*nums)[right + 1]){
                        right--
                    }else{
                        break
                    }
                }
            }
            
        }
    }
    return
}

func threeSum(nums []int) [][]int {
    res := [][]int{}
    sort.Ints(nums)
    
    if(len(nums) < 3){
        return res
    }
    
    if(nums[0] > 0){
        return res
    }
    
    helper(&nums, &res)
    
    return res
}